autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,570 @@
+    /// outputs where the input space is small enough.
+    fn possible_input_output_pairs(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Vec<SymbolicInputOutputPair<T, V>> {
+        let unknown_field_indices = bus_interaction
+            .payload
+            .iter()
+            .enumerate()
+            .filter(|(_i, expr)| expr.try_to_number().is_none())
+            .map(|(i, _)| i)
+            .collect::<Vec<_>>();
+        let expressions = unknown_field_indices
+            .iter()
+            .map(|i| (*i, bus_interaction.payload[*i].clone()))
+            .collect::<BTreeMap<_, _>>();
+        let range_constraints = expressions
+            .iter()
+            .map(|(i, expr)| (*i, self.solver.range_constraint_for_expression(expr)))
+            .collect::<BTreeMap<_, _>>();
+
+        // Currently, we only have hypotheses for:
+        // - 2 unknown fields (1 input, 1 output)
+        // - 3 unknown fields (2 inputs, 1 output)
+        if !(unknown_field_indices.len() == 2 || unknown_field_indices.len() == 3) {
+            return Vec::new();
+        }
+
+        unknown_field_indices
+            .clone()
+            .into_iter()
+            .combinations(unknown_field_indices.len() - 1)
+            .filter(|input_indices| {
+                self.has_small_domain(input_indices.iter().map(|i| range_constraints[i].clone()))
+            })
+            // Build the input-output pairs
+            .map(|input_indices| {
+                let output_index = unknown_field_indices
+                    .iter()
+                    .cloned()
+                    .filter(|i| !input_indices.contains(i))
+                    .exactly_one()
+                    .unwrap();
+                SymbolicInputOutputPair {
+                    inputs: input_indices
+                        .iter()
+                        .map(|i| SymbolicInput {
+                            index: *i,
+                            expression: expressions[i].clone(),
+                            range_constraint: range_constraints[i].clone(),
+                        })
+                        .collect(),
+                    output: SymbolicOutput {
+                        index: output_index,
+                        expression: expressions[&output_index].clone(),
+                    },
+                }
+            })
+            .collect()
+    }
+
+    /// Given a list of input range constraints, computes whether the total input space (i.e.,
+    /// the cross product of all possible input values) is small enough.
+    fn has_small_domain(
+        &self,
+        input_range_constraints: impl Iterator<Item = RangeConstraint<T>>,
+    ) -> bool {
+        input_range_constraints
+            .map(|rc| {
+                // TODO: This should share code with `has_few_possible_assignments`,
+                // But this only currently only considers the range width which ignores the mask
```

Comment:
We should split the `range_width` function into two: One that returns the exact range width and another that returns an upper bound on the number of possible values.